src/os/linux/mod.rs

```diff
@@ -0,0 +1,165 @@
+mod xdg;
+
+use std::{
+    ffi::{OsStr, OsString},
+    fmt::{self, Display},
+    path::{Path, PathBuf},
+};
+
+#[derive(Debug)]
+pub enum DetectEditorError {
+    NoDefaultEditorSet,
+    FreeDesktopEntryNotFound,
+    FreeDesktopEntryParseError,
+}
+
+impl Display for DetectEditorError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::NoDefaultEditorSet => write!(f, "No default editor is set: xdg-mime queries for \"text/rust\" and \"text/plain\" both failed"),
+            Self::FreeDesktopEntryNotFound => write!(f, "Entry Not Found: xdg-mime returned an entry name that could not be found"),
+            Self::FreeDesktopEntryParseError => write!(f, "Entry Parse Error: xdg-mime returned an entry that could not be parsed"),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum OpenFileError {
+    LaunchFailed(bossy::Error),
+}
+
+impl Display for OpenFileError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::LaunchFailed(e) => write!(f, "Launch failed: {}", e),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Application {
+    exec_command: OsString,
+    icon: Option<OsString>,
+    xdg_entry_path: PathBuf,
+}
+
+impl Application {
+    pub fn detect_editor() -> Result<Self, DetectEditorError> {
+        let entry = {
```

Comment:
This can be expressed as
```rust
let entry = xdg::query_mime_entry("text/rust")
    .or_else(|| xdg::query_mime_entry("text/plain"))
    .ok_or(DetectEditorError::NoDefaultEditorSet)?;
```